sru: 19.4.33 manual 4 test for multiple ssh keys (#92)
cloud-init/ubuntu-sru: bugs/sru-19.4.33-manual-4.txt

=== Begin SRU Template ===
[Impact]
Images which configure multiple ssh authorized key files for users via
setting AuthorizedKeysFiles = <file1> <file2> would result in the creation of
of an empty .ssh/ subdirectory.

In this scenario, old/broken cloud-init would emit and empty
.ssh authorized keys file named .ssh/"<file1> "/<file2>. Instead of
aggregating all pubkeys from <file1> and <file2> and adding any
cloud-config keys into <file1>, we are left with <file1 + cloud-config keys>
and an empty .ssh/"<file1> "/<file2>.  

This version of cloud-init now reads both <file1> and <file2>, collects this
list of keys and augments it with anything written by #cloud-config user data.
The resulting set of keys is written into <file1>, which on Ubuntu is
.ssh/authorized_keys

[Test Case]

> Create an lxc instance with AuthorizedKeyFiles configured for mulitple files:

Reviewer: ```suggestion
Create an lxc instance with AuthorizedKeyFiles configured for multiple files:
```

+ grep cloud-init
  cloud-init
Get:1 http://archive.ubuntu.com/ubuntu xenial-proposed/main amd64 cloud-init all 19.4-33-gbb4131a2-0ubuntu1~16.04.1 [413 kB]
dpkg-preconfigure: unable to re-open stdin: No such file or directory
Preparing to unpack .../cloud-init_19.4-33-gbb4131a2-0ubuntu1~16.04.1_all.deb ...
Unpacking cloud-init (19.4-33-gbb4131a2-0ubuntu1~16.04.1) over (19.3-41-gc4735dd3-0ubuntu1~16.04.1) ...
Setting up cloud-init (19.4-33-gbb4131a2-0ubuntu1~16.04.1) ...
Leaving 'diversion of /etc/init/ureadahead.conf to /etc/init/ureadahead.conf.disabled by cloud-init'
+ lxc exec test-xenial -- bash /setup_ssh_multi_keys.sh
+ lxc exec test-xenial -- cloud-init clean --logs --reboot
+ sleep 5
+ lxc exec test-xenial -- cloud-init status --wait --long
...................+ echo --- Expect only authorized_keys file in .ssh ---
--- Expect only authorized_keys file in .ssh ---
+ lxc exec test-xenial -- find /home/ubuntu/.ssh
/home/ubuntu/.ssh
/home/ubuntu/.ssh/authorized_keys
+ echo --- Expect chad.smith, localkey and defaultkey in authorized_keys ---
--- Expect chad.smith, localkey and defaultkey in authorized_keys ---
+ lxc exec test-xenial -- cat /home/ubuntu/.ssh/authorized_keys

> ssh-rsa localkey

Reviewer: This looks like a failure to me. We're missing the LP keys and the defaultkey.

Author: Strange, I think there was a timing issue in the manual test, if I perform the following now on the same lxc that I haven't touched since the test was run, it contains all expected keys:
```
lxc exec test-xenial cat /home/ubuntu/.ssh/authorized_keys
ssh-rsa localkey
ssh-rsa defaultkey
ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDSL7uWGj8cgWyIOaspgKdVy0cKJ+UTjfv7jBOjG2H/GN8bJVXy72XAvnhM0dUM+CCs8FOf0YlPX+Frvz2hKInrmRhZVwRSL129PasD12MlI3l44u6IwS1o/W86Q+tkQYEljtqDOo0a+cOsaZkvUNzUyEXUwz/lmYa6G4hMKZH4NBj7nbAAF96wsMCoyNwbWryBnDYUr6wMbjRR1J9Pw7Xh7WRC73wy4Va2YuOgbD3V/5ZrFPLbWZW/7TFXVrql04QVbyei4aiFR5n//GvoqwQDNe58LmbzX/xvxyKJYdny2zXmdAhMxbrpFQsfpkJ9E/H5w0yOdSvnWbUoG5xNGoOB csmith@fringe # ssh-import-id lp:chad.smith

ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDvl3VfPjVXsXBsm6r2J+UneIMr4ZOJhQlXuBWTwzexbd/XugB3k5EXA18yyqjEVT+bApVwlxATY66drVUPBuZ2JMU1HuLOKhG6toZd7j042oV5b2TEvg0es9qxs9mtGzvMPf3mB3tBVY/ESall023M+J5JjGGSO4J3zM/9c+P3Hs7xyCjAoySZDN2VZzscPgSGZzck8xtyO39uPfscKXi9LJkkhDDG6SVWie5OeM8TxyH2W2eNDKeXid/qgdIxqRLSYiNnWpt9htI0SzahnFYtsw9VLkij+0cM29lBIGUr5AehN2Y6jetxODR3pZt4YqOiyC6D5NaEsVGKOb0zjIBBCso6mIseejlOwocSYUH21YnLDS2Mu31bHRmPjpRvMVTOFtnS2OkfOxYTyMNFZ5PH/a0/t3DGxZZqz74F+APxG1X0vsgSFA9yYzbBaY3fr3vNAEYsRMTeBIjF6Gx6QmX3/kw5KBid4t8qQCV4Z1l8UmWZu4qFYxV/Z0IYPZazgYy/1W0qfRm5AdvpDdH9XArIokwqe1E2Djp5/xWp4Z9dAINmfJvNZxiDJk7gQz+Hdka/1U/f3wQSds9OAjF+a94Lj+F9CmMrhpVEZG5OL8ysK4iwSOsDhW7iLeZw5AO7cVhDUWj53/p2FP4+zxin/tYkDhNTJF0Nhc2uLMLxRCOGrQ== csmith@uptown # ssh-import-id lp:chad.smith
```

Author: sleep 10 FTW ( I think we may have a timing issue with cloud-init status --wait on lxc that we'll need to sort at some point).... like --wait doesn't block if the results.json or status.json is completely empty possibly.

Author: I added a card to look into the lxc reboot race https://trello.com/c/Epf5kwpH/50-test-cloud-init-status-wait-across-on-lxc-reboot-for-race

=== Begin SRU Template ===
[Impact]

> Images which provide multiple ssh authorized key files for users via

Reviewer: ```suggestion
Images which configure multiple SSH authorized key files by
```

> setting AuthorizedKeysFiles = <file1> <file2> would result in an invalid ssh

Reviewer: It's not an invalid SSH configuration; the fact that it's valid and we don't handle it is the problem we're fixing.

configuration resulting in no remote ssh connectivity due to the absence of
any .ssh/authorized_keys file.

> In this scenario, old/broken cloud-init would emit

Reviewer: "subdirectory" is in this sentence twice, which suggests to me it's mid-edit. Can we clarify this wording, please?

a separate subdirectory that would represent the space-delimited string value
as a single subdirectory representing the full string instead of consolidating
the keys from the separate AuthorizedKeysFiles represented in the
/etc/ssh/sshd_config AuthorizedKeysFiles value.

[Test Case]

> Create an lxc instance with AuthorizedKeyFiles set to mulitple files.

Reviewer: ```suggestion
Create an LXD instance with `AuthorizedKeyFile` set to multiple files.
```

.ssh/authorized_keys will contain one pre-existing key.
/default-keys will contain a one pre-existing default ssh pubkey.
Launch an lxc vm providing a third ssh pubkey via ssh_import_id.
Ensure all three keys are present in .ssh/authorized_keys.

#!/bin/sh
set -x

# To be adapted to the SRU to test
SRU_SERIES="xenial bionic eoan"

cat > importme.yaml <<EOF
## template: jinja
#cloud-config
ssh_import_id : [chad.smith]
EOF

cat > setup_ssh_multi_keys.sh <<EOF

> sed -i 's/#AuthorizedKeysFile.*/AuthorizedKeysFile	%h\/\.ssh\/authorized_keys \/default_keys/' /etc/ssh/sshd_config

Reviewer: I find it much easier to read path substitutions if you use something other than `/` for the sed delimiter:

```suggestion
sed -i 's,#AuthorizedKeysFile.*,AuthorizedKeysFile	%h/.ssh/authorized_keys /default_keys,' /etc/ssh/sshd_config
```

Author: Nice, Thanks TIL I could replace that delimiter

rm -rf /home/ubuntu/.ssh/
mkdir /home/ubuntu/.ssh
echo "ssh-rsa localkey" > /home/ubuntu/.ssh/authorized_keys
sudo chown -R ubuntu:ubuntu /home/ubuntu/.ssh
echo "ssh-rsa key2" > /default_keys
EOF

cat > setup_proposed.sh <<EOF
#/bin/bash
mirror=http://archive.ubuntu.com/ubuntu
echo deb \$mirror \$(lsb_release -sc)-proposed main | tee /etc/apt/sources.list.d/proposed.list
apt-get update -q
apt-get install -qy cloud-init
EOF

for series in $SRU_SERIES; do
  lxc launch ubuntu-daily:$series test-$series -c user.user-data="$(cat importme.yaml)"
  sleep 5
  lxc exec test-$series -- cloud-init status --wait --long
  lxc file push setup_proposed.sh setup_ssh_multi_keys.sh test-$series/
  lxc exec test-$series -- bash /setup_ssh_multi_keys.sh
  lxc exec test-$series -- cloud-init clean --logs --reboot
  sleep 5
  lxc exec test-$series -- cloud-init status --wait --long
  echo '--- Expect funky subdir in .ssh ---'
  lxc exec test-$series -- find /home/ubuntu/.ssh
  echo '--- Expect only chad.smith and localkey in .ssh/authorized_keys ---'
  lxc exec test-$series -- wc -l /home/ubuntu/.ssh/authorized_keys

>   lxc exec test-$series -- grep -v chad.smith /home/ubuntu/.ssh/authorized_keys

Reviewer: Filtering out one of the lines makes this a lot harder to follow:

```suggestion
  lxc exec test-$series -- cat /home/ubuntu/.ssh/authorized_keys
```

  lxc exec test-$series -- bash /setup_proposed.sh | grep cloud-init
  lxc exec test-$series -- bash /setup_ssh_multi_keys.sh
  lxc exec test-$series -- cloud-init clean --logs --reboot
  sleep 5
  lxc exec test-$series -- cloud-init status --wait --long
  echo '--- Expect only authorized_keys file in .ssh ---'
  lxc exec test-$series -- find /home/ubuntu/.ssh
  echo '--- Expect chad.smith, localkey and defaultkey in authorized_keys ---'
  lxc exec test-$series -- wc -l /home/ubuntu/.ssh/authorized_keys

>   lxc exec test-$series -- grep -v chad.smith /home/ubuntu/.ssh/authorized_keys

Reviewer: ```suggestion
  lxc exec test-$series -- cat /home/ubuntu/.ssh/authorized_keys
```

done

> 

Reviewer: Lots of newlines.

[Regression Potential]

> If ssh key publishing fails on single AuthorizedKeyFiles setup, vms would be

Reviewer: I don't really know what this means.

+ grep cloud-init
  cloud-init
Get:1 http://archive.ubuntu.com/ubuntu xenial-proposed/main amd64 cloud-init all 19.4-33-gbb4131a2-0ubuntu1~16.04.1 [413 kB]
dpkg-preconfigure: unable to re-open stdin: No such file or directory
Preparing to unpack .../cloud-init_19.4-33-gbb4131a2-0ubuntu1~16.04.1_all.deb ...
Unpacking cloud-init (19.4-33-gbb4131a2-0ubuntu1~16.04.1) over (19.3-41-gc4735dd3-0ubuntu1~16.04.1) ...
Setting up cloud-init (19.4-33-gbb4131a2-0ubuntu1~16.04.1) ...
Leaving 'diversion of /etc/init/ureadahead.conf to /etc/init/ureadahead.conf.disabled by cloud-init'
+ lxc exec test-xenial -- bash /setup_ssh_multi_keys.sh
+ lxc exec test-xenial -- cloud-init clean --logs --reboot
+ sleep 5
+ lxc exec test-xenial -- cloud-init status --wait --long
....
status: done
time: Mon, 27 Jan 2020 19:33:48 +0000
detail:
DataSourceNoCloud [seed=/var/lib/cloud/seed/nocloud-net][dsmode=net]
+ echo --- Expect only authorized_keys file in .ssh ---
--- Expect only authorized_keys file in .ssh ---
+ lxc exec test-xenial -- find /home/ubuntu/.ssh
/home/ubuntu/.ssh
/home/ubuntu/.ssh/authorized_keys
+ echo --- Expect chad.smith, localkey and defaultkey in authorized_keys ---

> --- Expect chad.smith, localkey and defaultkey in authorized_keys ---

Reviewer: This shows "key2" as present, not any of the listed expected keys.
